File: 2019/2.js

const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

function check(str1,str2,mazeSize){
    const lydiaPath = {}
    let x = 0;
    let y = 0;
    for(let i = 0;i<str1.length;i++){
        let nextX,nextY;
        if(str1[i] == 'E'){
            nextX = x + 1;
            nextY = y;
        } else {
            nextX = x;
            nextY = y + 1;
        }
        lydiaPath[x*mazeSize+y] = nextX*mazeSize+nextY;
        x = nextX;
        y = nextY;
    }
    x = 0;
    y = 0;
    for(let i = 0;i<str2.length;i++){
        let nextX,nextY;
        if(str2[i] == 'E'){
            nextX = x + 1;
            nextY = y;
        } else {
            nextX = x;
            nextY = y + 1;
        }
        if(lydiaPath[x*mazeSize+y] == nextX*mazeSize+nextY){
            console.log({x,y});
            return false;
        }
        x = nextX;
        y = nextY;
    }
    return true;
}


function proc(index,str){
    let ret = '';
    for(let i = 0;i<str.length;i++){
        if(str[i]=='E'){
            ret += 'S';
        } else {
            ret += 'E';
        }
    }
    console.log('Case #%d %s',index,ret);
}

function test(){
    let size = 50000;
    let eNum = 0;
    let sNum = 0;
    let str='';
    for(let i = 0;i<2*size-2;i++){
        if(Math.random() > 0.5){
            if(eNum < size-1){
                str += 'E';
                eNum ++;
            } else {
                str += 'S';
                sNum ++;
            }
        } else {
            if(sNum < size-1){
                str += 'S';
                sNum ++;
            } else {
                str += 'E';
                eNum ++;
            }
        }
    }
    proc(0,str);
}
let n;
let index=0;
let mazeSize = 0;
rl.on('line', function (line) {
    if(n == null){
        index = 0;
        mazeSize = 0;
        n = parseInt(line);
    } else if(n >= index) {
        if(mazeSize == 0){
            mazeSize = parseInt(line);
        } else {
            index += 1;
            proc(index,line);
            mazeSize = 0;
        }
    } else {
        process.exit();
    }
});
/*
for(let i = 0;i<1000;i++){
    test();
}
console.log('over');
*/

File: 2019/2.cpp
#include<iostream>
#include<map>
#include<vector>

using namespace std;

bool loop(int x,int y,vector<char> *s, const int size, const int index, map<int,int>* lydiaPath, int eNum, int sNum){
    if(x == size-1 && y == size-1){
        string str(s->begin(), s->end());
        cout << "Case #"<<index<<": "<<str<<endl;
        return true;
    }
    map<int,int>::iterator it = lydiaPath->find(x*size+y);
    int nextIndex = -1;
    if(it != lydiaPath->end()){
        nextIndex = it->second;
    }
    if(x < size-1 && nextIndex!= (x+1)*size+y && eNum < size-1){
        s->push_back('E');
        eNum ++;
        if(loop(x+1,y,s,size,index,lydiaPath,eNum,sNum)){
            return true;
        }
        eNum --;
        s->pop_back();
    }
    if(y < size-1 && nextIndex!= x*size+y+1 && sNum < size -1){
        s->push_back('S');
        sNum ++;
        if(loop(x,y+1,s,size,index,lydiaPath,eNum,sNum)){
            return true;
        }
        sNum --;
        s->pop_back();
    }
    return false;
}

void proc(int index,int size, string s){
    map<int,int> *lydiaPath = new map<int,int>();
    int x = 0;
    int y = 0;
    for(int i = 0;i<s.length();i++){
        int nextX = x;
        int nextY = y;
        if(s[i] == 'E'){
            nextX = x+1;
        } else {
            nextY = y+1;
        }
        lydiaPath->insert(pair<int,int>(x*size+y,nextX*size+nextY));
        x = nextX;
        y = nextY;
    }
    vector<char> *vec = new vector<char>(size*2-2);
    loop(0,0,vec,size,index,lydiaPath,0,0);
    delete lydiaPath;
    delete vec;
    
}

int test(){
    int size = 10000;
    char *chs = new char[2*size-1];
    chs[2*size-2] = '\0';
    srand(time(nullptr)); // 以当前时间为随机生成器的种子
    int eNum = 0;
    int sNum = 0;
    for(int i = 0; i< 2*size-2;i++){
        int random = rand() % 2;
        if(random == 0){
            if(eNum < size-1){
                chs[i] = 'E';
                eNum ++;
            } else {
                chs[i] = 'S';
                sNum ++;
            }
        } else{
            if(sNum < size-1){
                chs[i] = 'S';
                sNum ++;
            } else {
                chs[i] = 'E';
                eNum ++;
            }
        }
    }
    string *s = new string(chs);
    int start = time(nullptr);
    cout << "start:"<<start<<endl;
    proc(0,size,*s);
    int end = time(nullptr);
    cout << "end:"<<end<<endl;
    delete s;
    cout<< "Cost:"<<end-start<<"s"<<endl;
    return 0;
}

int main(){
    /*int t;
    cin >> t;
    for(int i=0;i<t;i++){
        int n;
        string s;
        cin >> n;
        cin >> s;
        proc(i+1,n,s);
    }*/
    test();
    return 0;
}